prov/shm/src/smr_ep.c

```diff
@@ -1337,266 +1193,34 @@ static void smr_init_ipc_socket(struct smr_ep *ep)
-	int ret;
-
-	srx_ctx = srx->ep_fid.fid.context;
-	ofi_spin_lock(&srx_ctx->lock);
-
-	match_attr.id = addr;
-	match_attr.tag = tag;
-
-	dlist_entry = dlist_find_first_match(&srx_ctx->trecv_queue.list,
-					     srx_ctx->trecv_queue.match_func,
-					     &match_attr);
-	if (!dlist_entry) {
-		smr_entry = smr_alloc_rx_entry(srx_ctx);
-		if (!smr_entry) {
-			ret = -FI_ENOMEM;
-		} else {
-			smr_entry->peer_entry.owner_context = NULL;
-			smr_entry->peer_entry.addr = addr;
-			smr_entry->peer_entry.size = size;
-			smr_entry->peer_entry.tag = tag;
-			smr_entry->peer_entry.srx = srx;
-			*rx_entry = &smr_entry->peer_entry;
-			ret = -FI_ENOENT;
-		}
-		goto out;
-	}
-	dlist_remove(dlist_entry);
-
-	*rx_entry = (struct fi_peer_rx_entry *) dlist_entry;
-	(*rx_entry)->srx = srx;
-	ret = FI_SUCCESS;
-out:
-	ofi_spin_unlock(&srx_ctx->lock);
-	return ret;
-}
-
-static int smr_queue_msg(struct fi_peer_rx_entry *rx_entry)
-{
-	struct smr_srx_ctx *srx_ctx = rx_entry->srx->ep_fid.fid.context;
-
-	ofi_spin_lock(&srx_ctx->lock);
-	dlist_insert_tail((struct dlist_entry *) rx_entry,
-			  &srx_ctx->unexp_msg_queue.list);
-	ofi_spin_unlock(&srx_ctx->lock);
-	return 0;
-}
-
-static int smr_queue_tag(struct fi_peer_rx_entry *rx_entry)
-{
-	struct smr_srx_ctx *srx_ctx = rx_entry->srx->ep_fid.fid.context;
-
-	ofi_spin_lock(&srx_ctx->lock);
-	dlist_insert_tail((struct dlist_entry *) rx_entry,
-			  &srx_ctx->unexp_tagged_queue.list);
-	ofi_spin_unlock(&srx_ctx->lock);
-	return 0;
-}
-
-static void smr_free_entry(struct fi_peer_rx_entry *entry)
+static int smr_discard(struct fi_peer_rx_entry *rx_entry)
 {
-	struct smr_srx_ctx *srx = (struct smr_srx_ctx *) entry->srx->ep_fid.fid.context;
-	struct smr_rx_entry *smr_entry, *owner_entry;
-
-	ofi_spin_lock(&srx->lock);
-	smr_entry = container_of(entry, struct smr_rx_entry, peer_entry);
-	if (entry->owner_context) {
-		owner_entry = container_of(entry->owner_context,
-					   struct smr_rx_entry, peer_entry);
-		if (!--owner_entry->multi_recv_ref &&
-		    owner_entry->peer_entry.size < srx->min_multi_recv_size) {
-			if (ofi_peer_cq_write(srx->cq,
-					      owner_entry->peer_entry.context,
-					      FI_MULTI_RECV, 0, NULL, 0, 0,
-					      FI_ADDR_NOTAVAIL)) {
-				FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-					"unable to write rx MULTI_RECV completion\n");
-			}
-			ofi_freestack_push(srx->recv_fs, owner_entry);
-		}
-	}
+	struct smr_cmd_ctx *cmd_ctx = rx_entry->peer_context;
 
-	ofi_freestack_push(srx->recv_fs, smr_entry);
-	ofi_spin_unlock(&srx->lock);
+	ofi_buf_free(cmd_ctx);
+	return FI_SUCCESS;
 }
 
-static struct fi_ops_srx_owner smr_srx_owner_ops = {
-	.size = sizeof(struct fi_ops_srx_owner),
-	.get_msg = smr_get_msg,
-	.get_tag = smr_get_tag,
-	.queue_msg = smr_queue_msg,
-	.queue_tag = smr_queue_tag,
-	.free_entry = smr_free_entry,
-};
-
-static int smr_discard(struct fi_peer_rx_entry *rx_entry)
+int smr_progress_addr(struct fi_peer_rx_entry *rx_entry)
 {
 	struct smr_cmd_ctx *cmd_ctx = rx_entry->peer_context;
```

Comment:
nit allocating cmd_ctx seems a little redundant

Reply:
What do you mean?

Reply:
I meant this is a one-line function
```
rx_entry->addr =
              rx_entry->peer_context->ep->region->map->peers[cmd_ctx->cmd.msg.hdr.id].fiaddr;
```

Also just noticed the signature
```
static void smr_get_addr(...)
```
IMO either it should return the address, or we call it 
```
static void smr_update_addr(...)
```

Reply:
Ah I understand what you mean. It's just a pointer, not allocation, and I thought setting it helped with readability since the line is already quite a doozy... and if I used it inline I would need a cast so the whole line would look like this:
```
rx_entry->addr = rx_entry->peer_context->ep->region->map->peers[((struct smr_cmd_ctx *) rx_entry->peer_context)->cmd.msg.hdr.id].fiaddr;
```

I get your issue with the update/get_addr naming. I feel pretty neutral about the names but I'll let Sean chime in about preferences.

@shefty, do you have any naming preferences for the API or return/setting?
Options could be to rename the owner op to update_addrs() to make it clearer that we are updating a list and/or set the peer op to update_addr() as well to indicate it will be changing the value.
Or get_addr could return the addr and the owner could set it to the return?

Reply:
> quite a doozy

Indeed. Didn't realize the array index was also that long. Then let's keep it as is.